### ISSUE TITLE
__new__ Class Method And Class Call Return Types

### ISSUE THREAD

**Issue by Daniihh** [microsoft/pyright#625]
PyRight doesn't take note of a class's class method `__new__`'s return type, causing class call's return types to be inaccurate.

Reproducing
-----------------
Create a class like so...
```py
class NotSoClass:
  def __new__(cls):
    return "Not the instance you were expecting!"
```
...then write the following...
`string = NotSoClass()`
Python sets the string variable with a value of the type `str`, while PyRight sets it's type to `NotSoClass`.

Expected Behavior
-------------------------
That PyRight would have knowledge of the `__new__` class method, and be able to adapt class call's return types to `__new__`'s return type.

Environment
-----------------
PyRight VSCode extension 1.1.32
Visual Studio Code 1.43.2
Linux

**Comment by erictraut** [comment]
This is intended behavior.

A type checker needs to make certain assumptions about what constitutes legitimate behavior and what does not. The code above violates one of the core assumptions in pyright. I verified that mypy makes the same assumption.

Out of curiosity, why would one ever want to do something as shown in the code sample above?

**Comment by Daniihh** [comment]
I was replacing the `__new__` function to make a higher order class. I wanted a class object that would construct another class, instead of construct an object that acted like a class.

**Comment by erictraut** [comment]
You can do that with a regular function rather than abusing the "class" construct.

**Comment by jordmantheman** [comment]
The one place I've found this to be useful is when using declarative Factories (e.g. factory-boy) for testing declarative frameworks (e.g. django).

```
import factory
from django.db import models

class Foo(models.Model):
    """model."""

class SomeFooFactory(factory.Factory):
    """ factory."""

    class Meta:
        model = Foo

    # I'd expect this to override any typing otherwise determined
    def __new__(cls, *args, **kwargs) -> Foo:
        return super().__new__(*args, **kwargs)
```

There are workarounds...

```
foo: Foo = SomeFooFactory() # or call create/stub directly.
```

🤷‍♂️  Just wanted to add some color to this scenario, as it's a pretty popular testing strategy and a pretty natural way to express things.
